core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -284,6 +291,32 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
     cachedSerializedStatuses.contains(shuffleId) || mapStatuses.contains(shuffleId)
   }
 
+  // Return the list of locations and blockSizes for each reducer.
+  // The map is keyed by reducerId and for each reducer the value contains the array
+  // of (location, size) of map outputs.
+  //
+  // This method is not thread-safe
+  def getStatusByReducer(
+      shuffleId: Int,
+      numReducers: Int)
```

Comment:
Also, what about changing this to something like "getMapLocationsWithBiggestOutputs", and having it also accept the reduceId (and then return Option[Array[(BlockManagerId, Long)]], since the caller always just takes one element out of the map? And then you can just add a comment internally saying that you populate the whole map for use later.